File: java/leetcode/src/main/java/com/leetcode/BFS_DFS/Combination_Sum_39.java
package com.leetcode.BFS_DFS;

import java.util.ArrayList;
import java.util.List;

public class Combination_Sum_39 {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> path = new ArrayList<>();
        dfs(candidates, 0, 0, target, path);
        return res;
    }

    void dfs(int[] candidates, int start, int sum, int target, List<Integer> path) {
        if (sum == target) {
            res.add(new ArrayList<Integer>(path));
        } else {
            for (int i = start; i < candidates.length; i++) {
                if(sum + candidates[i] > target) continue;
                path.add(candidates[i]);
                dfs(candidates, i, sum + candidates[i], target, path);
                path.remove(path.size() - 1);
            }
        }
    }
}

File: js/algrithm/huahua130/bfs-dfs/Combination_Sum_39.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const ans  = []
    const dfs  = function(start, sum, path) {
        if (sum === target) {
            ans.push(path.slice())
        } else {
            for (let i = start; i < candidates.length; i++) {
                const can = candidates[i]
                if (can + sum > target) continue
                path.push(can)
                dfs(i, can + sum, path)
                path.pop()
            }
        }

    }

    dfs(0, 0, [])

    return ans
};